src/quest/resources.py

```diff
@@ -0,0 +1,85 @@
+            return self
+
+        async def __aexit__(self, exc_type, exc_value, traceback):
+            self._on_close()
+            logging.debug(f'Resource stream closed for {id(self)}')
+
+    class ResourceStream:
+        def __init__(self,
+                     get_resources: Callable[[], Coroutine],
+                     is_workflow_complete: Callable[[], bool],
+                     on_close: Callable[['ResourceStreamManager.ResourceStream'], None]
+                     ):
+            self._stream_gate = asyncio.Event()
+            self._update_event = asyncio.Event()
+            self._get_resources = get_resources
+            self._is_workflow_complete = is_workflow_complete
+            self._on_close = on_close
+            # TODO - set any gates?
+
+        async def __aiter__(self):
+            """
+            Provide a stream of resource snapshots for this workflow.
+            Everytime the workflow resource state changes, an update will be published.
+
+            NOTE: Once you start the resource stream,
+            the workflow will not progress unless you iterate this stream.
+            Use wait_for_completion() to automatically iterate (and ignore) the remaining updates.
+            """
+
+            # The caller of this function lives outside the step management of the historian
+            #         -> don't replay, just yield event changes in realtime
+
+            async with ResourceStreamManager.ResourceStreamContext(lambda: self._on_close(self)):
+                # Yield the current resources immediately
+                yield await self._get_resources()
+
+                while not self._is_workflow_complete():
+                    # Yield new resources updates as they become available
+                    await self._update_event.wait()
+                    self._update_event.clear()
+                    yield await self._get_resources()
+                    self._stream_gate.set()
+
+    def get_resource_stream(self,
+                            identity,
+                            get_resources: Callable[[], Coroutine],
+                            is_workflow_complete: Callable[[], bool]
+                            ):
+        rs = ResourceStreamManager.ResourceStream(
+            get_resources,
+            is_workflow_complete,
+            lambda res_stream: self.resource_streams[identity].remove(res_stream)
+        )
+        if identity not in self.resource_streams:
+            self.resource_streams[identity] = set()
+        self.resource_streams[identity].add(rs)
+        return rs
+
+    async def update(self, identity):
+        # If there is no resources stream associated with `identity`, no update needed.
+        if identity not in self.resource_streams:
+            return
+
+        # TODO: Say a stream is added but then removed, the identity still exists as a key but just has an empty set.
+        # Is it okay to leave it as such?
```

Comment:
I think we should remove empty sets as we go. Otherwise we get a memory leak as we keep track of more and more empty sets over time.